design/tla/Task.tla

```diff
@@ -0,0 +1,760 @@
+
+   XXX: ``The updater takes care of making sure that each slot converges to having a single running task.''
+   How does it do that?
+   *)
+ReleaseReady ==
+  /\ UNCHANGED << services, nEvents >>
+  /\ \E id \in DOMAIN tasks :
+       LET sid  == tasks[id].service
+           slot == VSlot(id)
+       IN
+       /\ tasks[id].desired_state = ready   \* (and not e.g. `remove')
+       /\ State(id) = ready
+       /\ \A other \in TasksOfVSlot(sid, slot) \ {id} :
+             running \prec State(other)   \* All other tasks have finished
+       /\ tasks' = [tasks EXCEPT ![id].desired_state = running]
+
+(** XXX: `orchestrators.md' says it also ``deletes the linked tasks when a service is deleted.''
+    How do you delete a task? Presumably it needs to be shut down first, somehow. *)
+
+(* Tasks that the orchestrator must always do eventually if it can: *)
+OrchestratorProgress ==
+  \/ CreateSlot
+  \/ ReplaceTask
+  \/ RequestRemoval
+  \/ CleanupTerminated
+  \/ ReleaseReady
+
+(* All actions that the orchestrator can perform *)
+Orchestrator ==
+  \/ OrchestratorProgress
+  \/ RestartTask
+
+-------------------------------------------------------------------------------
+\*  Actions performed the allocator
+
+(* Pick a `new' task and move it to `pending'.
+
+   The spec says the allocator will ``allocate resources such as network attachments
+   which are necessary for the tasks to run''. However, we don't model any resources here. *)
+Allocator ==
+  /\ UNCHANGED << services, nEvents >>
+  /\ \E id \in DOMAIN tasks :
+     /\ State(id) = new
+     /\ tasks' = [tasks EXCEPT ![id].status.state = pending]
+
+-------------------------------------------------------------------------------
+\*  Actions performed the scheduler
+
+(* The scheduler assigns a node to a `pending' task and moves it to `assigned'
+   once sufficient resources are available (we don't model resources here). *)
+Scheduler ==
+  /\ UNCHANGED << services, nEvents >>
+  /\ \E id \in DOMAIN tasks :
+     /\ State(id) = pending
+     /\ LET requestedNode  == tasks[id].node
+            candidateNodes == IF requestedNode = unassigned
+                                THEN Node  \* (all nodes)
+                                ELSE { requestedNode }
+        IN
+        \E node \in candidateNodes :
+           tasks' = [tasks EXCEPT
+                        ![id].status.state = assigned,
+                        ![id].node = node
+                    ]
+
+-------------------------------------------------------------------------------
+\*  Actions performed by worker nodes (actually, by the dispatcher on their behalf)
+
+(* Try to advance containers towards `desired_state' if we're not there yet.
+
+   XXX: Spec says ``A task will progress through the ACCEPTED, PREPARING, and STARTING states on the way to RUNNING.''
+        I'm assuming the `ready' state occurs in between `preparing' and `starting'. *)
```

Comment:
I think it's for restarting tasks: the restart manager sets the desired state of the new task to `ready` and only moves it to `running` when the old task has finished shutting down. Perhaps it shouldn't even start shutting down the old task until the new one is ready? Also, the docs say it might wait a bit if there is a restart delay configured.